### PR TITLE
Restart spark interpreter when the spark context has shutdown

### PR DESCRIPTION
**Pull request by akhiluanandh**
When an interpreter is loaded from `NoteInterpreterLoader`, if this is a `SparkInterpreter` and the spark context has terminated (say, due to `sc.stop()` being called), this interpreter is useless. Right now, the user needs to go to the interpreter settings and restart this group before it can be used. Restarting the spark interpreter creates a new spark context.

In this patch, we do this automatically. When we load a spark interpreter and the spark context has stopped, we automatically restart it. We also give the user an option whether or not to enable this automatic restart or not (default true).

To check whether restart is required, we perform a thrift RPC call `restartRequired`. All interpreters except SparkInterpreter return false. SparkInterpreter returns `restartEnabled && sc != null && sc.taskScheduler() == null`, which is true if automatic restart is enabled and spark context has been stopped (stopping a spark context makes its taskScheduler null).

### CODE REVIEW

**Comment by djoelz** [comment]
@akhiluanandh could you please add unit tests to your changes? Especially to remoteinterpreterservice. Thanks a lot!

File: spark/src/main/java/org/apache/zeppelin/spark/SparkInterpreter.java [hunk 1]
```diff
@@ -109,6 +109,8 @@
             .add("zeppelin.spark.maxResult",
                 getSystemDefault("ZEPPELIN_SPARK_MAXRESULT", "zeppelin.spark.maxResult", "1000"),
                 "Max number of SparkSQL result to display.")
+            .add("spark.auto.restart.sc", "true", "spark context will automatically restart" +
```
Comment (felixcheung) [inline spark/src/main/java/org/apache/zeppelin/spark/SparkInterpreter.java]: If this is a zeppelin config, I think this should start with 'zeppelin', ie, zeppelin.spark.auto.restart.sc

Comment (felixcheung) [inline spark/src/main/java/org/apache/zeppelin/spark/SparkInterpreter.java]: it would be consistent with other zeppelin spark interpreter group settings

**Comment by Leemoonsoo** [comment]
@akhiluanandh Thanks for the contribution.

Can SparkInterpreter handle restart SparkContext locally without adding a new method in Interpreter class? ie. restart SparkContext only, not whole interpreter process.

Reasons i think restart only SparkContext is is better,

 * We can keep Interpreter class simple. One less thing to understand to implement new interpreter for Zeppelin.
 * When dependency loader loads some libraries, restarting SparkContext can reuse loaded libraries, but restarting interpreter process can not use these libraries, in current architecture. So user may have different libraries loaded in this environment after restart and it can be very confusing, especially when restarting is done silently.

What do you think @akhiluanandh ?

**Comment by felixcheung** [comment]
We could call `sc.stop()` or `SparkIMain.reset()`?
Though apparently SparkIMain reset() has some issue: https://issues.apache.org/jira/browse/SPARK-10039

**Comment by akhiluanandh** [comment]
Thanks for the comments everyone.
I agree that it would probably be better to restart only the SparkContext and not the interpreter. I will try to implement that.

**Comment by corneadoug** [comment]
@akhiluanandh 
I'm going around old PRs.
Any plan on continuing that one?

**Comment by akhiluanandh** [comment]
@corneadoug 
No plans to continue it in the near future. I'll close this PR.

**Pull request closed without merging**